### ISSUE TITLE
10.5beta1 PHP 7.4 - Cannot install apps from Market - Archives of type application/gzip are not supported - fileinfo magic

### ISSUE THREAD

**Issue by fmkaiser** [owncloud/core#37467]
### Steps to reproduce
```
> php7.4 occ market:install activity   [or any other app]
activity: Archives of type application/gzip are not supported 
```

The same error is also displayed when trying to install a market app via the web interface.

### Server configuration
**ownCloud version:** 10.5.0beta1  
**PHP version:** 7.4  
**Operating system 1**: Ubuntu 20.04 w/ (default) PHP 7.4
**Operating system 2**: CentOS 8 w/ PHP 7.4 from Remi repo

### Preliminary investigation
The error is generated in `lib/private/Installer.php:281`. It expects a mime type of `application/x-gzip`. With PHP 7.4, fileinfo magic seems to have been updated and now returns mime type `application/gzip` for gzipped files.

The change seems to have been [made upstream](https://github.com/file/file/commit/3c12eb8d1418181af93abdd466c64dcddf68ce16) in `file` and now [shipped with  PHP 7.4](https://github.com/php/php-src/commit/622b10f06e5f9ba080b748b7b5b1807b142f27fd). 

The different mime type returned by fileinfo can be easily reproduced with this [small PHP script](https://gist.github.com/fmkaiser/97403438bb5a923bf561336d8c3fddbd):
```
ubuntu2004:~# php7.4 finfo.php 
application/gzip; charset=binary
ubuntu2004:~# php7.3 finfo.php 
application/x-gzip; charset=binary
```

### Preliminary fix

```
diff --git a/lib/private/Installer.php b/lib/private/Installer.php
index 164c8bdf04..1126bd62d8 100644
--- a/lib/private/Installer.php
+++ b/lib/private/Installer.php
@@ -281,7 +281,7 @@ class Installer {
 
                //detect the archive type
                $mime = \OC::$server->getMimeTypeDetector()->detect($path);
-               if ($mime !=='application/zip' && $mime !== 'application/x-gzip' && $mime !== 'application/x-bzip2') {
+               if ($mime !=='application/zip' && $mime !== 'application/x-gzip' && $mime !== 'application/gzip' && $mime !== 'application/x-bzip2') {
                        throw new \Exception($l->t("Archives of type %s are not supported", [$mime]));
                }

diff --git a/lib/private/Archive/Archive.php b/lib/private/Archive/Archive.php
index be51cf5170..0b0378903b 100644
--- a/lib/private/Archive/Archive.php
+++ b/lib/private/Archive/Archive.php
@@ -46,6 +46,8 @@ abstract class Archive {
                                return new ZIP($path);
                        case 'application/x-gzip':
                                return new TAR($path);
+                       case 'application/gzip':
+                               return new TAR($path);
                        case 'application/x-bzip2':
                                return new TAR($path);
                }
```

The changes above allowed me to successfully run `occ market:install`. 

I am however not that familiar with the ownCloud codebase and have almost certainly missed some places that might need to be updated as well.

**Comment by shaobin** [comment]
I had the same issue and solved the problem similarly. 

> 
> ### Steps to reproduce
> 
> ```
> > php7.4 occ market:install activity   [or any other app]
> activity: Archives of type application/gzip are not supported 
> ```
> 
> The same error is also displayed when trying to install a market app via the web interface.
> ### Server configuration
> 
> **ownCloud version:** 10.5.0beta1
> **PHP version:** 7.4
> **Operating system 1**: Ubuntu 20.04 w/ (default) PHP 7.4
> **Operating system 2**: CentOS 8 w/ PHP 7.4 from Remi repo
> ### Preliminary investigation
> 
> The error is generated in `lib/private/Installer.php:281`. It expects a mime type of `application/x-gzip`. With PHP 7.4, fileinfo magic seems to have been updated and now returns mime type `application/gzip` for gzipped files.
> 
> The change seems to have been [made upstream](https://github.com/file/file/commit/3c12eb8d1418181af93abdd466c64dcddf68ce16) in `file` and now [shipped with PHP 7.4](https://github.com/php/php-src/commit/622b10f06e5f9ba080b748b7b5b1807b142f27fd).
> 
> The different mime type returned by fileinfo can be easily reproduced with this [small PHP script](https://gist.github.com/fmkaiser/97403438bb5a923bf561336d8c3fddbd):
> 
> ```
> ubuntu2004:~# php7.4 finfo.php 
> application/gzip; charset=binary
> ubuntu2004:~# php7.3 finfo.php 
> application/x-gzip; charset=binary
> ```
> 
> ### Preliminary fix
> 
> ```
> diff --git a/lib/private/Installer.php b/lib/private/Installer.php
> index 164c8bdf04..1126bd62d8 100644
> --- a/lib/private/Installer.php
> +++ b/lib/private/Installer.php
> @@ -281,7 +281,7 @@ class Installer {
>  
>                 //detect the archive type
>                 $mime = \OC::$server->getMimeTypeDetector()->detect($path);
> -               if ($mime !=='application/zip' && $mime !== 'application/x-gzip' && $mime !== 'application/x-bzip2') {
> +               if ($mime !=='application/zip' && $mime !== 'application/x-gzip' && $mime !== 'application/gzip' && $mime !== 'application/x-bzip2') {
>                         throw new \Exception($l->t("Archives of type %s are not supported", [$mime]));
>                 }
> 
> diff --git a/lib/private/Archive/Archive.php b/lib/private/Archive/Archive.php
> index be51cf5170..0b0378903b 100644
> --- a/lib/private/Archive/Archive.php
> +++ b/lib/private/Archive/Archive.php
> @@ -46,6 +46,8 @@ abstract class Archive {
>                                 return new ZIP($path);
>                         case 'application/x-gzip':
>                                 return new TAR($path);
> +                       case 'application/gzip':
> +                               return new TAR($path);
>                         case 'application/x-bzip2':
>                                 return new TAR($path);
>                 }
> ```
> 
> The changes above allowed me to successfully run `occ market:install`.
> 
> I am however not that familiar with the ownCloud codebase and have almost certainly missed some places that might need to be updated as well.

**Comment by phil-davis** [comment]
I get the same in a development environment:
```
$ php --version
PHP 7.4.7 (cli) (built: Jun 12 2020 07:44:05) ( NTS )
Copyright (c) The PHP Group
Zend Engine v3.4.0, Copyright (c) Zend Technologies
    with Zend OPcache v7.4.7, Copyright (c), by Zend Technologies
    with Xdebug v2.9.6, Copyright (c) 2002-2020, by Derick Rethans
$ php occ market:install activity
activity: Installing new app ...
 activity: Archives of type application/gzip are not supported 
```
And after the suggested code fix:
```
$ php occ market:install activity
activity: Installing new app ...
activity: App installed.
```

I raised PR #37559 We will see if reviewers have other suggestions. @fmkaiser thanks for the report and suggested fix.
